app/controllers/search_controller.rb

```diff
@@ -1,36 +1,18 @@
-# app/controllers/search_controller.rb
-require_relative '../services/google_search_service'
-
 class SearchController < ApplicationController
   def index
     if params[:query].present?
```

Comment:
I think I see the issue here. 

When you first hit the search result, you are enqueueing the `SearchJob` with `perform_later` and then returning the job id. (This is good exactly what I would expect)

However, because the job was enqueued, it hasn't yet run yet. Now when you click the second time, it goes into the `if params[:job_id]` branch below and then that will return the results. 

Because of the background job aspect, there is no real escaping that you will need to hit the controller a second time. The general pattern though is that instead of the user having to click, you would poll. This is typically done on the javascript side and you would use something like [setInterval](https://developer.mozilla.org/en-US/docs/Web/API/setInterval) to make the request to the `SearchController` automatically. 

To give you a sense of what this might look like, in Javascript, it might look like this (note: I haven't tested this code so it might not be fully correct)

```
    var refreshIntervalId;
    refreshIntervalId = setInterval(function () {
        console.log($.ajax('/search.json?job_id=<somejobid>'));
        clearInterval(refreshIntervalId); // This tells it to stop making the request every 500ms. Call this when you get the successful results
    } , 500);     
```

This will log the JSON results to the console (you'd probably want to put it onto the page) and it will make the result every 500 milliseconds

Reply:
I was thinking about creating websockets or channels to deal with that, instead of having JS request every 500ms. From the data I read, our BingSearch API takes 400ms to return. That could lead to almost 1 second waiting time.

Following this implementation, our next step will involve enriching the search results. We plan to pass the data through an AI API with a customized prompt, which will then return enhanced results. This approach should give us a decent USP while raising the bar for cashflow-positive competitors.

I tried to build a channel, but couldn't make it work properly(I could do it with 2 clicks). Maybe it's the issue of good_job gem, I will try to use sidekiq. So we can work on advancing that together tonight.

Reply:
Another step would be to set up 'rack/attack' to create 2 scopes of limits. 

1) 3 requests per second.
2) 1000 requests per month.

To achieve that I was planning to store the data in our Redis service, so we could persist it while also deploying new versions of our app.

The app is online (search currently w/o background jobs): 
- Landing page: https://ecosearch-test.oasisofchange.org/
- Search Feature: https://ecosearch-test.oasisofchange.org/search

Reply:
The idea of using webscokets to push the data back to the client side is good. That is one alternative to polling. 

That said, it is also a little more complicated. I do suggest to stick with a simple polling approach for now until you have everything working, and then work on swapping the polling for websockets. 

I don't think there is anything `good_job` related that is really impacting this behavior. This is just the nature of background jobs, and you will likely see the same issues with sidekiq

---

Rack attack would also not be super useful here. Rack attack is used for throttling this will stop requests from happening. This is very useful to create some limits so you don't make thousands of requests per second and you can protect the system from users who would act in ways that don't fit what you need.

Reply:
Thank you for the feedback and today's session.